### PR TITLE
[GSProcessing] Add Docker-related files and code utilities.

### PR DESCRIPTION
**Pull request by thvasilo**
*Issue #, if available:*

*Description of changes:*
* Add Docker-related files
* Add code utilities used by `DistHeterogeneousGraphLoader`.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: graphstorm-processing/graphstorm_processing/graph_loaders/schema_utils.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+    Raises
+    ------
+    NotImplementedError
+        In case an unsupported feature_type is provided.
+    """
+    # TODO: Replace with pattern matching after moving to Python 3.10?
+    if feature_type in [
+        "category",
+        "datetime",
```
Comment (classicsong) [inline graphstorm-processing/graphstorm_processing/graph_loaders/schema_utils.py:98]: Do we support datetime, multi_category and multi_numerical now?

File: graphstorm-processing/graphstorm_processing/graph_loaders/schema_utils.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+
+def _parse_features_schema(features_objects: Sequence[FeatureConfig]) -> Sequence[StructField]:
+    field_list = []
+    for feature_config in features_objects:
+        feature_type = feature_config.feat_type
+        for feature_col, _ in zip(feature_config.cols, feature_config.feat_name):
+            spark_feature_type = determine_spark_feature_type(feature_type)
+
+            field_list.append(StructField(feature_col, spark_feature_type(), True))
```
Comment (classicsong) [inline graphstorm-processing/graphstorm_processing/graph_loaders/schema_utils.py:72]: Shall we also put feature name into the StructField?

File: graphstorm-processing/graphstorm_processing/graph_loaders/schema_utils.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    Returns
+    -------
+    StructType
+        A Spark StructType describing the typed schema for the edge files.
+    """
+    # TODO: Will need to add support for relation types
+    edge_fields_list = [
+        StructField(edge_config.src_col, StringType(), False),
+        StructField(edge_config.dst_col, StringType(), False),
+        # StructField(relation_type, StringType(), False),
```
Comment (classicsong) [inline graphstorm-processing/graphstorm_processing/graph_loaders/schema_utils.py:52]: Can you remove this?

File: graphstorm-processing/graphstorm_processing/graph_loaders/row_count_utils.py [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+                        "do not match row counts of structure files for corresponding type %s:\n%s",
+                        feature_name,
+                        feature_counts,
+                        type_name,
+                        structure_counts,
+                    )
+        return all_match
+
+    @staticmethod
+    def verify_all_features_match(data_meta: Dict[str, Dict]) -> bool:
```
[the remark this answers is not in the document]
Comment (thvasilo) [reply graphstorm-processing/graphstorm_processing/graph_loaders/row_count_utils.py:318]: There's a difference in the semantics between the two functions, in `verify_features_and_structure_match` we verify counts match across feature type and corresponding structure types and this is only possible for edges (there's no need to have this verification for node types).

However, there is the assumption that all features for each node type match, so we treat features separately and we check features for edges and nodes separately in `add_row_counts_to_metadata` rows 90-91 in this PR revision.

File: graphstorm-processing/graphstorm_processing/graph_loaders/schema_utils.py (hunk 2 above)
Comment (thvasilo) [reply graphstorm-processing/graphstorm_processing/graph_loaders/schema_utils.py:72]: This is about describing a single column of the Spark dataframe when reading it so it needs to have the column name. When saving the output we name the columns by the feature names.

File: graphstorm-processing/graphstorm_processing/graph_loaders/schema_utils.py (hunk 1 above)
Comment (thvasilo) [reply graphstorm-processing/graphstorm_processing/graph_loaders/schema_utils.py:98]: We don't I'll remove these.

File: graphstorm-processing/docker/0.1.0/Dockerfile.cpu [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+# Starts framework
+ENTRYPOINT ["bash", "/usr/lib/spark/code/docker-entry.sh"]
+
+FROM base AS prod
+RUN python3 -m pip install /usr/lib/spark/code/graphstorm_processing-*.whl && \
+    rm /usr/lib/spark/code/graphstorm_processing-*.whl
+CMD ["gs-processing"]
+
+FROM base AS test
+RUN python3 -m pip install /usr/lib/spark/code/graphstorm-processing/
```
[the remark this answers is not in the document]
Comment (classicsong) [reply graphstorm-processing/docker/0.1.0/Dockerfile.cpu:44]: Prod for production? make sense

File: graphstorm-processing/docker/build_gsprocessing_image.sh [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+msg "- VERSION: ${VERSION}"
+
+# Prepare Docker build directory
+rm -rf "${BUILD_DIR}/docker/code"
+mkdir -p "${BUILD_DIR}/docker/code"
+
+if [[ ${TARGET} == "prod" ]]; then
+    # Build the graphstorm-processing library and copy the wheels file
+    poetry build -C ${GSP_HOME} --format wheel
+    cp ${GSP_HOME}/dist/graphstorm_processing-${VERSION}-py3-none-any.whl \
```
[the remark this answers is not in the document]
Comment (classicsong) [reply graphstorm-processing/docker/build_gsprocessing_image.sh:114]: I am thinking about releasing and hosting an official .whl for 0.2 and later releases.

File: graphstorm-processing/graphstorm_processing/graph_loaders/row_count_utils.py [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+        self.output_prefix = output_prefix
+        self.filesystem_type = filesystem_type
+        self.metadata_dict = metadata_dict
+        if self.filesystem_type == "s3":
+            output_bucket = self.output_prefix.split("/")[2]
+            bucket_region = s3_utils.get_bucket_region(output_bucket)
+            # Increase default retries because we are likely to run into
+            # throttling errors
+            self.pyarrow_fs = fs.S3FileSystem(
+                region=bucket_region, retry_strategy=fs.AwsStandardS3RetryStrategy(max_attempts=10)
```
[the remark this answers is not in the document]
Comment (classicsong) [reply graphstorm-processing/graphstorm_processing/graph_loaders/row_count_utils.py:55]: If PyArrow will report the error. There will be fine.

File: graphstorm-processing/graphstorm_processing/graph_loaders/row_count_utils.py [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+        relative_parquet_file_path : str
+            A file path relative to the `output_prefix` the object was initialized with.
+
+        Returns
+        -------
+        int
+            The number of rows in the Parquet file.
+        """
+        if self.filesystem_type == "s3":
+            file_s3_uri = f"{self.output_prefix}/{relative_parquet_file_path}"
```
[the remark this answers is not in the document]
Comment (classicsong) [reply graphstorm-processing/graphstorm_processing/graph_loaders/row_count_utils.py:127]: we can have three functions, this do make the code more readable

File: graphstorm-processing/graphstorm_processing/graph_loaders/row_count_utils.py [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+        Returns
+        -------
+        Sequence[int]
+            A list of file counts that corresponds to the order of the file paths
+            passed in.
+        """
+        # TODO: Despite parallel call this can still be slow for thousands of files.
+        # See if we can skip or at least do fully async
+        cpu_count = os.cpu_count() if os.cpu_count() else 4
+        assert cpu_count
```
[the remark this answers is not in the document]
Comment (classicsong) [reply graphstorm-processing/graphstorm_processing/graph_loaders/row_count_utils.py:162]: I C. Can you add a comment here about this? We actually tricks the mypy

File: graphstorm-processing/graphstorm_processing/graph_loaders/row_count_utils.py [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+        row_counts_per_file = Parallel(n_jobs=min(16, cpu_count), backend="threading")(
+            delayed(self.get_rows_for_parquet_file)(parquet_path)
+            for parquet_path in parquet_file_paths
+        )
+
+        return row_counts_per_file
+
+    def _add_counts_for_structure(
```
[the remark this answers is not in the document]
Comment (classicsong) [reply graphstorm-processing/graphstorm_processing/graph_loaders/row_count_utils.py:170]: Maybe explicitly say graph_structure. This make the code more readable.

File: graphstorm-processing/graphstorm_processing/graph_loaders/row_count_utils.py [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+                    type_name,
+                    top_level_key,
+                    self.metadata_dict[top_level_key].keys(),
+                )
+                continue
+            for feature_name, feature_data_dict in self.metadata_dict[top_level_key][
+                type_name
+            ].items():
+                relative_file_list = feature_data_dict["data"]  # type: Sequence[str]
+                logging.info(
```
[the remark this answers is not in the document]
Comment (classicsong) [reply graphstorm-processing/graphstorm_processing/graph_loaders/row_count_utils.py:252]: Yes. We can make it a debug logging.

File: graphstorm-processing/graphstorm_processing/graph_loaders/row_count_utils.py [hunk 12]
```diff
@@ -0,0 +1,368 @@
...
+                    "Getting counts for %s, type: %s, feature: %s",
+                    top_level_key,
+                    type_name,
+                    feature_name,
+                )
+                feature_row_counts = self.get_rows_for_parquet_files(relative_file_list)
+                feature_data_dict["row_counts"] = feature_row_counts
+                features_per_type_counts.append(feature_row_counts)
```
[the remark this answers is not in the document]
Comment (classicsong) [reply graphstorm-processing/graphstorm_processing/graph_loaders/row_count_utils.py:260]: Can you add some comments about this here?

File: graphstorm-processing/graphstorm_processing/graph_loaders/row_count_utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (classicsong) [reply graphstorm-processing/graphstorm_processing/graph_loaders/row_count_utils.py:348]: I mean in verify_features_and_structure_match we can call verify_all_features_match to verify features and call another logic to verify structures

File: graphstorm-processing/graphstorm_processing/graph_loaders/row_count_utils.py [hunk 13]
```diff
@@ -170,37 +170,69 @@ def get_rows_for_parquet_files(self, parquet_file_paths: Sequence[str]) -> Seque
...
+    def _add_counts_for_features(self, top_level_key: str, edge_or_node_type_key: str) -> None:
+        """Returns a nested list of counts for each feature, either for edges features
+        or node features. Modifies `self.metadata_dict` in place.
+
+        Parameters
+        ----------
+        top_level_key : str
+            The top level key that refers to the features we'll be getting
+            counts for, can be "edge_data" to get counts for edge features,
+            or "node_type" to get counts for node features.
```
Comment (classicsong) [inline graphstorm-processing/graphstorm_processing/graph_loaders/row_count_utils.py:219]: Why not node_data? (to make it consistent with edge_data.

File: graphstorm-processing/docker/README.md
Comment (classicsong) [inline graphstorm-processing/docker/README.md:1]: @zhjwy9343 Can you take a look of this README?

Comment (thvasilo) [reply graphstorm-processing/docker/README.md:1]: The doc referred to in the README is this: https://github.com/thvasilo/graphstorm/blob/gs-processing/graphstorm-processing/docs/source/usage/distributed-processing-setup.rst (will need some changes once this PR is merged)

File: graphstorm-processing/graphstorm_processing/graph_loaders/row_count_utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (thvasilo) [reply graphstorm-processing/graphstorm_processing/graph_loaders/row_count_utils.py:348]: Maybe I'm misunderstanding something, my guess is you'd like us to call `verify_all_features_match` within `verify_features_and_structure_match`? 

The iteration logic between matching features with structure is quite different from just matching features which would require a rewrite, I'd rather leave this as is unless you think it's a blocker. 

That leaves each function with a single purpose as described by their names: One is to ensure edges graph structures match with edge features, the other is to ensure feature types match with each other.

**Review by classicsong — APPROVED** [review]

**Pull request merged**